pkg/core/consensus/initiator/initiator.go

```diff
@@ -0,0 +1,122 @@
+
+func LaunchConsensus(eventBroker wire.EventBroker, rpcBus *wire.RPCBus, w *wallet.Wallet, counter *chainsync.Counter) {
+	// TODO: sync first
+	go startProvisioner(eventBroker, rpcBus, w, counter)
+	go startBlockGenerator(eventBroker, rpcBus, w)
+}
+
+func startProvisioner(eventBroker wire.EventBroker, rpcBus *wire.RPCBus, w *wallet.Wallet, counter *chainsync.Counter) {
+	// Setting up the consensus factory
+	f := factory.New(eventBroker, rpcBus, config.ConsensusTimeOut, w.ConsensusKeys())
+	f.StartConsensus()
+
+	// Get current height
+	req := wire.NewRequest(bytes.Buffer{}, 1)
+	resultBuf, err := rpcBus.Call(wire.GetLastBlock, req)
+	if err != nil {
+		l.WithError(err).Warnln("could not retrieve current height, starting from 1")
+		sendInitMessage(eventBroker, 1)
+		return
+	}
+
+	var currentHeight uint64
+	if err := encoding.ReadUint64(&resultBuf, binary.LittleEndian, &currentHeight); err != nil {
+		l.WithError(err).Warnln("could not decode current height, starting from 1")
+		sendInitMessage(eventBroker, 1)
+		return
+	}
+
+	if currentHeight > 0 {
+		// Get starting round
+		startingRound := getStartingRound(eventBroker, counter)
+		sendInitMessage(eventBroker, startingRound)
+		return
+	}
+
+	// We are at genesis. Start from 1
+	sendInitMessage(eventBroker, 1)
+}
+
+func startBlockGenerator(eventBroker wire.EventBroker, rpcBus *wire.RPCBus, w *wallet.Wallet) {
+	// make some random keys to sign the seed with
+	keys, err := user.NewRandKeys()
+	if err != nil {
+		l.WithError(err).Warnln("could not start block generation component - problem generating keys")
+		return
+	}
+
+	// reconstruct k
+	k, err := w.ReconstructK()
+	if err != nil {
+		l.WithError(err).Warnln("could not start block generation component - problem reconstructing K")
+		return
+	}
+
+	// get public key that the rewards should go to
+	publicKey := w.PublicKey()
+
+	// launch generation component
+	go func() {
+		if err := generation.Launch(eventBroker, rpcBus, k, keys, &publicKey, nil, nil, nil); err != nil {
+			l.WithError(err).Warnln("error launching block generation component")
+		}
+	}()
+}
+
+func getStartingRound(eventBroker wire.EventBroker, counter *chainsync.Counter) uint64 {
+	// Start listening for accepted blocks, regardless of if we found stakes or not
+	acceptedBlockChan, listener := consensus.InitAcceptedBlockUpdate(eventBroker)
+	// Unsubscribe from AcceptedBlock once we're done
+	defer listener.Quit()
+
+	// Sync first
+	syncToTip(acceptedBlockChan, counter)
+
+	for {
+		blk := <-acceptedBlockChan
+		return blk.Header.Height + 1
+	}
+}
+
+func sendInitMessage(publisher wire.EventPublisher, startingRound uint64) {
+	roundBytes := make([]byte, 8)
+	binary.LittleEndian.PutUint64(roundBytes, startingRound)
+	publisher.Publish(msg.InitializationTopic, bytes.NewBuffer(roundBytes))
+}
+
+func syncToTip(acceptedBlockChan <-chan block.Block, counter *chainsync.Counter) {
+	i := 0
+	for {
+		<-acceptedBlockChan
+		if counter.IsSyncing() {
```

Comment:
Theoretically, once syncing procedure reaches 500 blocks or time-out, IsSyncing would return false and syncToTip could exit without reaching the network tip.

Would this affect the consensus?

Reply:
> syncToTip could exit without reaching the network tip

It shouldn't, as it would need multiple blocks accepted in a row without triggering the sync counter before being able to exit.

```
     i++
     if i > 2 {
       break
     }
```

Reply:
Ah, good point. Makes sense.
One more question, so that means anytime the node is restarted and the wallet is loaded, it will always need to wait for at least 3 blocks accepted before initializing the consensus?

Reply:
It would need to wait for 3 accepted blocks in a row, where `IsSyncing()` returns false. Then we assume we are at the tip, and we start consensus.